consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,434 @@
+		return "precommit"
+	default:
+		return "unknown"
+	}
+}
+
+const (
+	maxFutureHeight = height(5)
+	maxFutureRound  = round(5)
+)
+
+type timeoutFn func(r round) time.Duration
+
+type Addr interface {
+	// Ethereum Addresses are 20 bytes
+	~[20]byte | felt.Felt
+}
+
+type Hash interface {
+	~[32]byte | felt.Felt
+}
+
+// Hashable's Hash() is used as ID()
+type Hashable[H Hash] interface {
+	Hash() H
+}
+
+type Application[V Hashable[H], H Hash] interface {
+	// Value returns the value to the Tendermint consensus algorith which can be proposed to other validators.
+	Value() V
+
+	// Valid returns true if the provided value is valid according to the application context.
+	Valid(V) bool
+}
+
+type Blockchain[V Hashable[H], H Hash, A Addr] interface {
+	// Height return the current blockchain height
+	Height() height
+
+	// Commit is called by Tendermint when a block has been decided on and can be committed to the DB.
+	Commit(height, V, []Precommit[H, A])
+}
+
+type Validators[A Addr] interface {
+	// TotalVotingPower represents N which is required to calculate the thresholds.
+	TotalVotingPower(height) uint
+
+	// ValidatorVotingPower returns the voting power of the a single validator. This is also required to implement
+	// various thresholds. The assumption is that a single validator cannot have voting power more than f.
+	ValidatorVotingPower(A) uint
+
+	// Proposer returns the proposer of the current round and height.
+	Proposer(height, round) A
+}
+
+type Slasher[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
+	// Equivocation informs the slasher that a validator has sent conflicting messages. Thus it can decide whether to
+	// slash the validator and by how much.
+	Equivocation(msgs ...M)
+}
+
+type Listener[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
+	// Listen would return consensus messages to Tendermint which are set // by the validator set.
+	Listen() <-chan M
+}
+
+type Broadcaster[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
+	// Broadcast will broadcast the message to the whole validator set. The function should not be blocking.
+	Broadcast(M)
+
+	// SendMsg would send a message to a specific validator. This would be required for helping send resquest and
+	// response message to help a specifc validator to catch up.
+	SendMsg(A, M)
+}
+
+type Listeners[V Hashable[H], H Hash, A Addr] struct {
+	ProposalListener  Listener[Proposal[V, H, A], V, H, A]
+	PrevoteListener   Listener[Prevote[H, A], V, H, A]
+	PrecommitListener Listener[Precommit[H, A], V, H, A]
+}
+
+type Broadcasters[V Hashable[H], H Hash, A Addr] struct {
+	ProposalBroadcaster  Broadcaster[Proposal[V, H, A], V, H, A]
+	PrevoteBroadcaster   Broadcaster[Prevote[H, A], V, H, A]
+	PrecommitBroadcaster Broadcaster[Precommit[H, A], V, H, A]
+}
+
+type Tendermint[V Hashable[H], H Hash, A Addr] struct {
+	nodeAddr A
+
+	state state[V, H] // Todo: Does state need to be protected?
+
+	messages       messages[V, H, A]
```

Comment:
I'm understanding that currently we're iterating over all messages to calculate aggregates. I think this can be more performant by storing only the aggregates.

Reply:
As far as I understand, we can store:
- Proposed value by (height, round)
- Prevote counter and Precommit counter by (height, round) and by (height, round, ID)
Each counter consists of a total count and set of nodes sending it. We can even improve this part by replace Address as key by node index in current instance and use bit array instead of set.

Reply:
Yes, these are certainly some optimisations which can be made, for the time being, I was aiming for simplicity. So having q or f is not the only thing which is important, the sender of the message is important too. So depending on how one designs the map there are different trades off. We probably need to think about our access patterns.